resolver/resolvers.go

```diff
@@ -200,10 +212,162 @@ func createResolver(resolverURL, source string) (*Resolver, bool, error) {
 		}
 	}
 
+	// Resolve ip if was not specfied by the user
+	err = checkAndResolveServerAddressAndPort(newResolver)
+	if err != nil {
+		return nil, false, err
+	}
+
 	newResolver.Conn = resolverConnFactory(newResolver)
 	return newResolver, false, nil
 }
 
+func checkURLParameterValidity(scheme string, isHostnameDomain bool, query url.Values) error {
+	for key := range query {
+		switch key {
+		case parameterName,
+			parameterVerify,
+			parameterIP,
+			parameterBlockedIf,
+			parameterSearch,
+			parameterSearchOnly,
+			parameterPath:
+			// Known key, continue.
+		default:
+			// Unknown key, abort.
+			return fmt.Errorf(`unknown parameter "%s"`, key)
+		}
+	}
+
+	verifyDomain := query.Get(parameterVerify)
+	paramterServerIP := query.Get(parameterIP)
+
+	if scheme == ServerTypeDoT || scheme == ServerTypeDoH {
+
+		switch {
+		case isHostnameDomain && verifyDomain != "":
+			return fmt.Errorf("cannot have verify parameter with a domain as a hostname")
+		case !isHostnameDomain && verifyDomain == "":
+			return fmt.Errorf("verify paremeter must be set when using ip as domain")
+		case !isHostnameDomain && paramterServerIP != "":
+			return fmt.Errorf("cannot have ip parameter while domain is an ip")
+		}
+	} else {
+		if verifyDomain != "" {
+			return fmt.Errorf("domain verification only supported in DoT and DoH")
+		}
+		if verifyDomain == "" && !isHostnameDomain {
+			return fmt.Errorf("DoT must have a verify query parameter set")
+		}
+	}
+
+	if scheme != ServerTypeDoH {
+		path := query.Get(parameterPath)
+		if path != "" {
+			return fmt.Errorf("path parameter is only supported in DoH")
+		}
+	}
+
+	return nil
+}
+
+func checkAndResolveServerAddressAndPort(resolver *Resolver) error {
+	if resolver.ServerAddress == "" {
+		resolverIps, err := resolveDomainIP(context.Background(), resolver.VerifyDomain)
```

Comment:
Instead of resolving here, just return `ErrContinue` for that domain in `Query`, then the next resolver will be used for that query and all the caching will be used too.